Add notebook sample to show the case of writing new asset

revit13 opened the pull request:
Closes #1459 

Following suggestions given in issue #1459 this sample shows three scenarios (a) how fybrik prevents writing new asset due to governance restrictions (suggested by @shlomitk1) (b) how to write data generated by the workload to an object store. (c) how to read data from a dataset stored in an object store

the sample is added to the site quick start.

This PR also created a notebook test for the write flow and converts the current notebook test to read flow notebook test.

simanadler commented on site/docs/samples/notebook-write.md:
```diff
@@ -0,0 +1,463 @@
+package dataapi.authz
+
+rule[{"action": {"name":"RedactAction","columns": column_names}, "policy": description}] {
+  description := "Redact written columns tagged as sensitive in datasets tagged with finance = true"
```
Description doesn't mention anything about neverland

revit13 replied:
Fixed. Thanks

simanadler commented on site/docs/samples/notebook-write.md:
````diff
@@ -0,0 +1,463 @@
+kubectl delete fybrikapplications.app.fybrik.io my-notebook-write -n fybrik-notebook-sample
+```
+
+## Scenario 3: Read the new written data
````
newly written data

revit13 replied:
Fixed. Thanks

simanadler commented on site/docs/samples/notebook-write.md:
```diff
@@ -0,0 +1,463 @@
+
+### Define data access policies to read the new data
+
+Define an [OpenPolicyAgent](https://www.openpolicyagent.org/) policy to redact the columns tagged as `PII` in datasets tagged with `finance` when reading data. Below is the policy (written in [Rego](https://www.openpolicyagent.org/docs/latest/policy-language/#what-is-rego) language):
```
So we are differentiating between "sensitive" which got blocked during the write flow, and PII which was written but will be blocked during the subsequent "read" flow?  How / by whom was was the PII tag added?

revit13 replied:
I changed the read scenario to not include redact actions as the focus is the writing. I also changed the written data to be from `/PS_20174392719_1491204439457_log.csv` as in the read flow.

revit13 commented on site/docs/samples/notebook-write.md:
````diff
@@ -0,0 +1,463 @@
+while [[ $(kubectl get fybrikapplication my-notebook-write -o 'jsonpath={.status.assetStates.new-data.conditions[?(@.type == "Ready")].status}') != "True" ]]; do echo "waiting for new-data asset" && sleep 5; done
+```
+
+Run the following command to extract the new asset id in the catalog. This asset id will be used in the third secnario when we try to read the new asset.
````
We take it from there: `.status.assetStates.new-data.catalogedAsset` I will rephrase the sentence to make it clearer. Thanks

revit13 commented on site/docs/samples/notebook-write.md:
```diff
@@ -0,0 +1,463 @@
+ENDPOINT_SCHEME=$(kubectl get fybrikapplication my-notebook-write -o jsonpath={.status.assetStates.new-data.endpoint.fybrik-arrow-flight.scheme})
+ENDPOINT_HOSTNAME=$(kubectl get fybrikapplication my-notebook-write -o jsonpath={.status.assetStates.new-data.endpoint.fybrik-arrow-flight.hostname})
+ENDPOINT_PORT=$(kubectl get fybrikapplication my-notebook-write -o jsonpath={.status.assetStates.new-data.endpoint.fybrik-arrow-flight.port})
+printf "${ENDPOINT_SCHEME}://${ENDPOINT_HOSTNAME}:${ENDPOINT_PORT}"
```
I think its simpler to copy-paste it. but if others think otherwise I can add script.

shlomitk1 commented on site/docs/samples/notebook-write.md:
```diff
@@ -0,0 +1,463 @@
+
+## Deploy resources for write scenarios
+
+Deploy [Datashim](https://github.com/datashim-io/datashim): 
```
Datashim and data-movement-operator have helm releases of their own. Same for vault that is required for Fybrik flows but is deployed separately.

revit13 commented on site/docs/samples/notebook-write.md:
````diff
@@ -0,0 +1,463 @@
+  ```
+  df_read
+  ```
+5. Execute all notebook cells and notice that the `Name` and `Email` columns appear redacted.
````
Fixed. Thanks

revit13 commented on site/docs/samples/notebook-write.md:
````diff
@@ -0,0 +1,463 @@
+```rego
+package dataapi.authz
+
+rule[{"policy": description}] {
````
I added a link. Thanks

simanadler replied:
You linked to rego, but perhaps it would be better to link to our page describing how to write opa data governance policies.  https://fybrik.io/dev/tasks/using-opa/

revit13 commented on site/docs/samples/notebook-write.md:
```diff
@@ -0,0 +1,463 @@
+* The `selector` field matches the labels of our Jupyter notebook workload.
+* The `data` field includes a `dataSetID` that matches the asset identifier in the catalog.
+* The `protocol` indicates that the developer wants to consume the data using Apache Arrow Flight. For some protocols a `dataformat` can be specified as well (e.g., `s3` protocol and `parquet` format).
+* The `isNewDataSet` field indicates we are writing a new asset.
```
Fixed. Thanks

revit13 commented on site/docs/samples/notebook-write.md:
```diff
@@ -0,0 +1,463 @@
+
+### Cleanup scenario one
+
+Before prceeding to scenarios two the OPA policy and fybrikapplications should be deleted:
```
Fixed. Thanks

revit13 commented on site/docs/samples/notebook-write.md:
````diff
@@ -0,0 +1,463 @@
+kubectl delete fybrikapplications.app.fybrik.io my-notebook-write -n fybrik-notebook-sample
+```
+
+Notice that `FybrikStorageAccount` resources are still appliedi after the cleanup.
````
Fixed. Thanks

revit13 commented on site/docs/samples/notebook-write.md:
````diff
@@ -0,0 +1,463 @@
+```
+
+Notice that `FybrikStorageAccount` resources are still appliedi after the cleanup.
+## Scenarios two: write new data
````
Fixed. Thanks

revit13 commented on site/docs/samples/notebook-write.md:
````diff
@@ -0,0 +1,463 @@
+```rego
+package dataapi.authz
+
+rule[{"action": {"name":"RedactAction","columns": column_names}, "policy": description}] {
````
Fixed. Thanks

revit13 commented on site/docs/samples/notebook-write.md:
```diff
@@ -0,0 +1,463 @@
+# Notebook sample for write flow
+
+This sample shows three scenarios (a) how fybrik prevents writing new asset due to governance restrictions (b) how to write data generated by the workload to an object store. (c) how to read data from a dataset stored in an object store
```
Fixed. thanks

revit13 commented on site/docs/samples/notebook-write.md:
````diff
@@ -0,0 +1,463 @@
+while [[ $(kubectl get fybrikapplication my-notebook-write -n fybrik-notebook-sample -o 'jsonpath={.status.assetStates.new-data.conditions[?(@.type == "Deny")].status}') != "True" ]]; do echo "waiting for my-notebook-write asset" && sleep 5; done
+```
+
+It is expected that the asset status is denied due to the policy defined above. Next, a new policy will be applied which will allow the writing to `theshire` object store.
````
The expression above in the `while` include those details...

revit13 commented:
@simanadler thanks for your comments! I changed the read scenario to not include redact actions as the focus is the writing. I also changed the written data to be from /PS_20174392719_1491204439457_log.csv as in the read flow.
In addition, I added test for the write flow as is described in this `.md` file and converted the current notebook sample test to include only the read flow as it appears in the site.

roytman commented on .github/workflows/build.yml:
```diff
@@ -145,16 +145,41 @@ jobs:
-      run: make run-notebook-tests
+      run: make run-notebook-readflow-tests
+
+ notebook-sample-write-flow:
```
pls fix identation

revit13 replied:
Fixed. Thanks!

simanadler commented on site/docs/samples/notebook-write.md:
```diff
@@ -0,0 +1,405 @@
+
+### Create a `FybrikApplication` resource for the notebook
+
+Re-apply `FybrikApplication` resource from scenarios one section.
```
Re-apply FybrikApplication as defined in scenario 1.

revit13 replied:
Fixed. Thanks

simanadler commented on site/mkdocs.yml:
```diff
@@ -11,7 +11,12 @@ nav:
   - get-started/quickstart.md
-  - samples/notebook.md
+  - Notebook samples:
+    - samples/tools.md
```
Perhaps this should be under an intro / before we begin?

revit13 replied:
I moved it under before we begin page

simanadler reviewed (commented):
Looks good.  See minor comments

revit13 commented on site/docs/samples/notebook-write.md:
```diff
@@ -0,0 +1,405 @@
+# write the new dataset
+file_path = "/path/to/PS_20174392719_1491204439457_log.csv"
+my_table = csv.read_csv(file_path)
+writer, _ = client.do_put(fl.FlightDescriptor.for_command(json.dumps(request)),
```
Added. Thanks

revit13 commented on site/docs/samples/notebook-write.md:
```diff
@@ -0,0 +1,405 @@
+
+In this sample a Jupyter notebook is used as the user workload and its business logic requires writing the new asset. Deploy a notebook to your cluster: execute the instructions from `Deploy a Jupyter notebook` section in the [notebook sample for the read flow](notebook-read.md) to deploy a Jupyter notebook.
+
+### Create a `FybrikApplication` resource for the notebook
```
Fixed. Thanks

revit13 commented on site/docs/samples/notebook-write.md:
```diff
@@ -0,0 +1,405 @@
+
+### Define data access policies for writing the data
+
+Define an [OpenPolicyAgent](https://www.openpolicyagent.org/) policy to return the list of column names tagged as sensitive whose destination is not neverland when the actionType is write. The columns are passed to the `FybrikModule` together with RedactAction upon deployment of the module by Fybrik.
```
Fixed. Thanks

revit13 commented on site/docs/samples/notebook-write.md:
```diff
@@ -0,0 +1,405 @@
+This sample shows three scenarios (a) how fybrik prevents writing a new asset due to governance restrictions (b) how to write data generated by the workload to an object store. (c) how to read data from a dataset stored in an object store
+
+In this sample you play multiple roles:
+- As a data steward you setup data governance policies 
```
Fixed. Thanks

revit13 commented on site/docs/samples/tools.md:
```diff
@@ -0,0 +1,7 @@
+# Tools used by the actors
+- The data owner would typically register the dataset in a proprietary or open source catalog.  We have provided [katalog](../reference/katalog.md) - a thin layer acting as a replacement for the data catalog for evaluation purposes. This simplifies the sample deployment. 
+- The data owner stores credentials for accessing the dataset in kubernetes secrets. 
+- Proprietary and open source data governance systems are available either as part of a data catalog or as stand-alone systems.  This sample uses the open source [OpenPolicyAgent](https://www.openpolicyagent.org/).  The data steward writes the policies in OPA's [rego](https://www.openpolicyagent.org/docs/latest/policy-language/#what-is-rego) language. 
```
Fixed. Thanks

revit13 commented on site/docs/samples/notebook-write.md:
```diff
@@ -0,0 +1,405 @@
+
+Note that for evaluation purposes the same object store is used for different regions in the storage accounts.
+
+## Scenarios one: write is forbidden due to governance restrictions
```
Fixed. Thanks

revit13 commented on site/docs/samples/notebook-write.md:
```diff
@@ -0,0 +1,405 @@
+
+## Scenarios one: write is forbidden due to governance restrictions
+
+### Define data access policies for write
```
Fixed. Thanks

revit13 commented on site/docs/samples/notebook-write.md:
```diff
@@ -0,0 +1,405 @@
+
+### Define data access policies for write
+
+Define an [OpenPolicyAgent](https://www.openpolicyagent.org/) policy to forbid to write sensitive data to regions `neverland` and `theshire` in datasets tagged with `finance`. This policy prevents the writing as the deployed fybrik storage account resources applied are in `neverland` and `theshire`.
```
Fixed. Thanks

revit13 commented on site/docs/samples/cleanup.md:
```diff
@@ -0,0 +1,13 @@
+# Cleanup
+
+When you’re finished experimenting with the notebook sample, clean it up:
```
Fixed. Thanks

revit13 commented on site/docs/samples/cleanup.md:
````diff
@@ -0,0 +1,13 @@
+    ```bash
+    kubectl delete namespace fybrik-notebook-sample
+    ```
+3. Delete the policy created on fybrik-system namespace:
````
Fixed. Thanks

revit13 commented on site/docs/samples/cleanup.md:
```diff
@@ -0,0 +1,13 @@
+# Cleanup
+
+When you’re finished experimenting with the notebook sample, clean it up:
```
Fixed. Thanks

revit13 commented on site/docs/samples/notebook-read.md:
```diff
@@ -1,40 +1,13 @@
 
 This sample shows how Fybrik enables a Jupyter notebook workload to access a cataloged dataset.
 It demonstrates how policies regarding the use of personal information are seamlessly applied when accessing a dataset containing financial data.
 
```
Done. Thanks
